src/raft/recv_install_snapshot.c

```diff
@@ -8,6 +8,350 @@
+ * back to the Leader and the leader puts incomming payloads of Signature()
+ * message into the HT.
+ *
+ * 4. When the follower sends the checksum of its highest numbered page to the
+ * Leader, it sends the SignatureResult() message using the done=true flag,
+ * upon receiving it the Leader moves into SNAPSHOT_INSTALLATION_STARTED state.
+ *
+ * 5. In SNAPSHOT_INSTALLATION_STARTED state, the Leader starts iterating over
+ * the local persistent state, and calculates the checksum for each page the
+ * state has. Then, it tries to find the checksum it calculated in HT. Based on
+ * the result of this calculation, the Leader sends InstallShapshot(CP..) or
+ * InstallShapshot(MV..) to the Follower.
+ *
+ * Upon receving these messages, the Follower moves into
+ * SNAPSHOT_CHUNCK_RECEIVED state. The Leader moves into SNAPSHOT_CHUNCK_SENT
+ * state after receiving first reply from the Follower.
+ *
+ * 6. When the iteration has finished the Leader sends
+ * InstallShapshot(..., done=true) message to the Follower. It moves the
+ * Follower back to NORMAL state and the state machine corresponding to the
+ * Follower on the Leader is moved to SNAPSHOT_DONE_SENT state.
+ *
+ * 7. The Leader sends AppendEntries() RPC to the Follower and restarts the
+ * algorithm from (1). The Leader's state machine is being moved to
+ * FOLLOWER_ONLINE state.
+ *
+ * =Failure model
+ *
+ * ==Unavailability of the Leader and Follower.
+ *
+ * To handle use-cases when any party of the communication becomes
+ * unavailable for a while without crash the following assumtions are
+ * made:
+ *
+ * - Signature() or InstallSnapshot(MV/CP) messages are idempotent and
+ *   can be applied to the persistent state many times resulting the
+ *   same transition.
+ *
+ * - Each message with data chuncks has an information about the
+ *   "chunk index". Chunk indexes come in monotonically increasing
+ *   order.
+ *
+ * - Each reply message acknowledges that the data received (or
+ *   ignored) by sending `result` field back to the counter part along
+ *   with last known chunk index as a confirmation that the receiver
+ *   "knows everything up to the given chunck index".
+ *
+ * - If a party notices that last known chunk index sent back to it
+ *   doesn't match it's own, the communication get's restarted from
+ *   the lowest known index.
+ *
+ * If a reply is not received the Leader will eventually timeout and retry
+ * sending the same message.
+ *
+ * ==Crashes of the Leader and Follower.
+ *
+ * Crashes of the Leader are handled by Raft when a new leader is elected
+ * and the snapshot process is restarted.
+ *
+ * If the Follower crashes, it will restart its state machine into the NORMAL
+ * state. If the Leader then sends a message which assumes the Follower is at
+ * the state prior to the crash, the Follower will reply using the message's
+ * result RPC using the failed=true flag. Upong receiving the message the
```

Comment:
I feel like I'm being picky to words, still it's better not to feed the reader of this DLD by unfulfilled hopes that the algorithm can detect a failure of the counter part of the communication. 

I'd propose to rename (sorry, it's not about words) `failed` flag with `unexpected` having a semantics that the Leader re-started the communication because the Follower by any reason treats incoming messages come "in unexpected order" w.r.t. defined states and expectations of the Follower (today it's just restarted, but tomorrow the same mechanism is good to use elsewhere, let's not limit ourselves).

> If the Leader then sends a message which assumes the Follower is at the state prior to the crash

Also, I'd propose to phrase quoted part of the sentence in terms of "unexpected sequence of events".

Reply:
Additional notes: 

Historically, my initial intent was to put a flag similar to unexpected into `AppendEntriesReply()` message so that it could be the only mechanism delivering such results. I'm not proposing to change anything in the currently provided mechanism (because I don't have a clear proposal by this time), still I'll be happy beyond any beliefs if we'll be able to generalize it.

Reply:
I don't think you are being picky with words, the DLD is very important and will later be read in order to understand the code so it is better to spend more time and do it properly. In https://github.com/canonical/dqlite/pull/657/commits/6633bd83a2cf2c99805d1acbcbf9224d6485dd5d I have changed the `failed` flag to `unexpected` and tried to reword the text with your suggestions. I am not sure I addressed all of your concerns so if there is anything missing please tell me.